### DISCUSSION TITLE
Need help in debugging the websocket connection

### DISCUSSION THREAD

**Discussion by var19** [digitallyinduced/ihp#1176] [Q&A] [answered]
So the issue is that sending a message is only working once, at the start. Any message after that isn't getting sent. 

_HTML_
```
<form onsubmit="sendMsg()">
  <textarea id="msg"></textarea>
  <button type="submit"> Send</button>
</form>
```
_JS_
```
  var controller = new WebSocket('.......');
  controller.onmessage = function (event) {
      console.log(event.data);
  };
  
  function sendMsg() {
      const msg = document.getElementById('msg').value;
      controller.send(msg);
  }
```

_Controller_
```
instance WSApp MyController where
    initialState = MyController
    run = do
        msg :: Text <- receiveData
        sendTextData (msg)
        forever receiveDataMessage
```

After a lot of debugging, I found that, for some reason `controller.send(msg)` is only working the first time. Everything apart from that is working properly. Any ideas why this could be happening?

**Comment by mpscholten** [comment, marked as answer]
>  Any message after that isn't getting sent.

What's the error?

If you want the websocket to output the received message in a loop you need this:

```haskell
instance WSApp MyController where
    initialState = MyController
    run = do
        forever do
            msg :: Text <- receiveData
            sendTextData (msg)
```

**Reply by var19** [reply]
Yes, this worked. Thank you.

However, the guide mentions using `forever receiveDataMessage`. Why does that not work?

**Reply by mpscholten** [reply]
Technically it should work. But practically it only keeps receiving data and then not doing anything with it. I'll update the docs to clarify this.

**Reply by var19** [reply]
Oh okay. That makes sense since there were no errors. It looked as if `controller.send(msg)` isn't working properly since the controller also stopped receiving data after the first message.
